compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1012,6 +1012,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         cause: Option<ObligationCause<'tcx>>,
     ) -> RelateResult<'tcx, Ty<'tcx>> {
         let source = self.try_structurally_resolve_type(expr.span, expr_ty);
+        let target = self.try_structurally_resolve_type(
```

Comment:
I take it back -- @lqd: this line may have changed the order we process obligations, re: perf #114604. 

I'd be surprised if this had a meaningful effect on perf, tho.

Reply:
I’ll check if cachegrind has anything interesting to say

Reply:
I'm also just getting rid of this one line and running perf in #114648.

Reply:
If this is the culprit, I guess I may be able to work around it.